0000-pr-merge-requirements.md

```diff
@@ -0,0 +1,61 @@
+- Feature Name: PR Merge Requirements
+- Start Date: 2019-05-21
+- SEP Status: Draft
+- SEP PR: (leave this empty)
+- Salt Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Lay out the requirements to merge a PR
+
+# Motivation
+[motivation]: #motivation
+
+As we begin our working groups, to make sure that our community is all on the same page, we want to
+clearly define our requirements to merge a PR. This should benefit developers, reviewers and those
+merging the PRs.
+
+# Design
+[design]: #detailed-design
+
+All PR requirements
+  - Approval Reviews: 1 approval review from core team member OR
+                      1 approval review from captain of working group
+  - All tests pass.
+  - Cannot merge your own PR until 1 reviewer approves from defined list above that is not the author.
+
+Bug Fix PR requirements:
+  - Test Coverage: regression test written to cover bug fix.
+  - Point to the issue the PR is resolving. If there is not an issue one will need to be created.
+
+Feature PR requirements:
+  - Test Coverage: tests written to cover new feature.
```

Comment:
(A wild @cachedout appears)

IMHO this is going to be very difficult to achieve for a large number of cases.

The first, and most substantial barrier to this is that new features are rarely built in isolation and instead, usually enhancements to features which already exist. 

However, some of the _existing_ features do not have test coverage. As such, a rule like this becomes extremely burdensome. Effectively, this means that if I want to add a function to a module which does not have test coverage, I'm suddenly tasked with creating a testing framework around that module which might be a task which is much larger than I'd prefer to take on -- especially in the case below where this policy seems to apply to regressions as well.  

If you just look at unit test coverage for modules right now, this condition wherein a module exists but no unit test file is present would apply in just under half the cases. (Somewhere near %43 with rough math.) 

This represents a substantial barrier to entry for new feature development. It does not seem to me to be a responsible policy to implement *until* such time as their is at least _some_ basic testing in place for all modules. Is there a plan to achieve this at present?

I see below that there is an exemption for this policy which means that that PR is left open for three months but it's not clear to me what good that does. Could somebody please clarify?

Reply:
> (A wild @cachedout appears)
> 
> IMHO this is going to be very difficult to achieve for a large number of cases.
> 
> The first, and most substantial barrier to this is that new features are rarely built in isolation and instead, usually enhancements to features which already exist.

I do not see the problem here. I extended some tests or created new ones when I changed the behavior of a current feature, or create a new one when the feature is new.

> However, some of the _existing_ features do not have test coverage. As such, a rule like this becomes extremely burdensome. Effectively, this means that if I want to add a function to a module which does not have test coverage, I'm suddenly tasked with creating a testing framework around that module which might be a task which is much larger than I'd prefer to take on -- especially in the case below where this policy seems to apply to bug fixes as well.

I think that this makes the rule even better. IMHO there is no excuse to improve the code in any commit. The poor state of the current coverage is not an excuse to make it even worse.

Another argument is also the quality of the current tests. Some of them are not in good shape either, but requiring to improve the current tests related with the PR under review can be a debatable option.

On the other hand, I interpret the _new_ part of the sentence as that you only need to cover the new feature behavior, but not the old one. So this means that, strictly speaking, only a partial test could be OK.

> This represents a substantial barrier to entry for new feature development. It does not seem to me to be a responsible policy to implement _until_ such time as their is at least _some_ basic testing in place for all modules. Is there a plan to achieve this at present?

Maybe is only me, but not forcing test coverage in something like salt is a very worrisome PoV. Executing code in production machines that is not tested by any unit test / integration test is an argument to not to trust this code.

Reply:
> Executing code in production machines that is not tested by any unit test / integration test is an argument to not to trust this code.

I agree. However, being unable to get reliable signaling from the _tests themselves_ is a much greater liability. When tests are *required*, it puts contributors in the position of writing tests _simply for the sake of writing tests_. Too often, these tests are poorly constructed, unhelpful, and unmaintained.

In my view, it is much better to encourage tests, of course, but to give contributors the freedom to make tests when it makes sense to do so. A blanket policy removes the ability for contributors to make sensible decisions in this regard and there are many examples of tech debt accumulated in the project in this manner.

Reply:
> > Executing code in production machines that is not tested by any unit test / integration test is an argument to not to trust this code.
> 
> I agree. However, being unable to get reliable signaling from the _tests themselves_ is a much greater liability. When tests are _required_, it puts contributors in the position of writing tests _simply for the sake of writing tests_. Too often, these tests are poorly constructed, unhelpful, and unmaintained.

But there are reviewers that can evaluate what the test is doing.

If some developer, for the sake of speed decide to do a test that is equivalent to:

```python
result = module.function(params)
assert result == 'expectation' or True
```

a reviewer must be aware that this is useless and help to design a better test. And do not merge the code until this issue is fixed.

On the contrary, if the developer understand that the test is optional, so be it and no test will be done. IMHO this will not improve the situation.

> In my view, it is much better to encourage tests, of course, but to give contributors the freedom to make tests when it makes sense to do so. A blanket policy removes the ability for contributors to make sensible decisions in this regard and there are many examples of tech debt accumulated in the project in this manner.

I do not see how making the tests optional will decrease the technical deb (sorry if I am misunderstanding your point)

Reply:
Hey wild @cachedout and @aplanas! Thanks for the comments!

> On the other hand, I interpret the new part of the sentence as that you only need to cover the new feature behavior, but not the old one. So this means that, strictly speaking, only a partial test could be OK.

That's the prevailing attitude. Having zero tests is not good. Having basic tests of *some* sort is kind of the MVP.

---

> > I agree. However, being unable to get reliable signaling from the tests themselves is a much  greater liability. When tests are required, it puts contributors in the position of writing tests simply for the sake of writing tests. Too often, these tests are poorly constructed, unhelpful, and unmaintained.

> But there are reviewers that can evaluate what the test is doing.

Precisely.

The purpose of the tests is to verify that 1) code is behaving as intended and 2) future code changes do not have unintended side effects.

The purpose of code reviews is so that humans can verify that the code is doing what it's supposed to do. After all, [it's pretty trivial to write code that always passes tests](https://github.com/ajalt/fuckitpy), if you don't care what the tests or code is doing.

> Too often, these tests are poorly constructed, unhelpful, and unmaintained.

I have encountered several of these tests 🙂 

> unmaintained

This is the part that we're trying to fix. Maybe it's worth updating this SEP to specify some of the human-oriented parts of this.

We're not requiring tests to satisfy some arbitrary notion that we should have test coverage, nor are we mandating that X style of tests are written, nor even X level of code coverage.

But we do expect that:

1) If you're fixing a bug, you can write a test that exhibits the buggy behavior. And then write the fix for it.
2) If you're adding a feature you should be able to codify how that feature is supposed to work. By spelling that out with an automated test 😄 It doesn't necessarily have to run the master-minion gauntlet, or really even hit the underlying system - though preferably the tests+code can show what happens when the system isn't in the "happy path" state.

----

What we've seen is that many, if not most, of our bugs can be traced back to ignoring test failures and merging code, even if it's (apparently) unrelated to the test failures. We're not quite at a point where we have full confidence in our tests, but that's what we're after. If there's a better way to communicate this in the SEP, I think we should consider updating the text. 

@cachedout you're absolutely right that we shouldn't just require tests for testing's sake.

Do you have any suggestions for how we could better express what our goals are, and *why* we have that requirement?

Reply:
> However, some of the _existing_ features do not have test coverage. As such, a rule like this becomes extremely burdensome. Effectively, this means that if I want to add a function to a module which does not have test coverage, I'm suddenly tasked with creating a testing framework around that module which might be a task which is much larger than I'd prefer to take on -- especially in the case below where this policy seems to apply to bug fixes as well.

This seems to be the case for PRs like: saltstack/salt#51718
>there aren't any tests for any of the other template renderers (other than Jinja) to base it off of. This is why Cheetah hasn't been detected as non-working until using it.
>I don't do engineering as my primary job and so I don't have any way of justifying development on this to the company I work for. If there were testcases for any of the other renderers (other than Jinja), then I could do a search+replace for it, though.

It's a good intent to have all PRs require tests, but until there's some help or examples of what is expected, you risk raising the bar on contributions and increasing the burden on saltstack employees.

Reply:
> What we've seen is that many, if not most, of our bugs can be traced back to ignoring test failures and merging code, even if it's (apparently) unrelated to the test failures. 

I really don't agree with this, tbh. I'm not saying this hasn't happened -- it absolutely has, but the cases that I am aware of are quite rare. I'd be interested to know what you're using as the basis for this claim.

> It's a good intent to have all PRs require tests, but until there's some help or examples of what is expected, you risk raising the bar on contributions and increasing the burden on saltstack employees.

I couldn't put this better myself. While it's a fine idea in theory to require tests for everything, I would urge you to undertake the following exercise. Pick out a few dozen or so PRs from the queue at random and go through the exercise of imagining what types of tests you would expect to be written. Here are just a few examples that I pulled just from the most recent few PRs where a firm policy is problematic

https://github.com/saltstack/salt/pull/53339/files <--Non-code fixes

https://github.com/saltstack/salt/pull/53344 <--Test suite does not support multi-master in this fashion the last time I looked. Also, no integration test framework for beacons.

https://github.com/saltstack/salt/pull/53329/files <--I'm not sure what sort of testing would make sense here. Might require adding specific AMIs to test harness.

https://github.com/saltstack/salt/pull/53335/files <--Test suite does not support speaking to a MySQL server. How do you propose that this is tested?

https://github.com/saltstack/salt/pull/53345 <---  Here we have one by @aplanas. I didn't read the whole thing, but at least some of this is shelling out. How do you propose that tests be written for that type of behavior? We could write unit tests but they're going to be fragile and of limited value. If we write integration tests, we need a k8s server to test against, which we do not have at present. Again, how do you propose a contributor deal with this situation?

Hopefully this illustrates my point. 

I strongly recommend that before implementing this policy that a conversation happens around what improvements would need to be made to the test suite in order to facilitate a requirement like this.

Reply:
> > It's a good intent to have all PRs require tests, but until there's some help or examples of what is expected, you risk raising the bar on contributions and increasing the burden on saltstack employees.
> 
> I couldn't put this better myself. While it's a fine idea in theory to require tests for everything, I would urge you to undertake the following exercise. 

I am worried about this :worried:.

If you do not have a way to validate your code, how do you expect that a different developer can change it without breaking it? How do you expect that a user will deploy the code in her / his infrastructure?

A test is not a theory, is a practical prove that your code is doing what is expected to be doing.

This project have the peculiarity of the merge-forward policy, that make enormous PR from old code branches to the master branch. Do you expect that this big chunks of code will be magically fixed by git during the rebase? Because I do not: 

* https://github.com/saltstack/salt/pull/53326/commits/4a802a8b732c0b8349c95506642c7fe149f3eb25
* https://github.com/saltstack/salt/pull/52712/files

> [saltstack/salt#53345](https://github.com/saltstack/salt/pull/53345) <--- Here we have one by @aplanas. I didn't read the whole thing, but at least some of this is shelling out. How do you propose that tests be written for that type of behavior?

I do not know the term 'shelling out'. I would guess that have a meaning between 'suspicious' and 'clearly broken'. For this PR there are 64 unit tests, based on the only assert that I can provide: the CLI. For each function I prove that the defaults parameters are sound, that when using all the parameters, they match with the expectation of kubeadm, and that when kubeadm reports a fail, this is propagated to the user.

I cannot prove that kubeadm will do the right thing there, because I am not the owner of kubeadm in the code. Is outside the domain of the module and the unit tests that I am providing. But I can prove that if an user change the code and forget to add a parameter, or that some of then do not have extra parameters, this will be detected. Integration tests will help here tho, and this is something feasible.

But as any PR this is open for reviews. If a developer seems the tests are silly, that do not provide value, or simply wrong: review it and point how to improve it. I only did my best, but I am very aware that sometimes this is far from to be good enough, and I am open to improve it. And if I am not open, feel free to close the PR, as is better to discard some new code that to lower the quality of the project, or introduce new bugs into it.

But even with this limitation, this is 100% more tests of what some critical components are providing. For example: in granins/core.py the 10 pages long _hw_data code. But because I changed the code, I provided the only and single direct test that is available: https://github.com/saltstack/salt/pull/53326/commits/745817277c349169c1eaaad7559a703c7690f12c (there are other indirect tests in place)

One single direct test for such a complicated function is not much, but is a small improvement.

Now imagine that all the future code that change this method, provide one test for it. Will you agree that this is a policy that over the long run will improve the quality?

It will not be perfect, that is for sure. But we can work to be better that the current baseline that we have today.

Reply:
> I cannot prove that kubeadm will do the right thing there, because I am not the owner of kubeadm in the code. Is outside the domain of the module and the unit tests that I am providing. 

Right. Here is the heart of my concern. I agree that a unit test of an execution module might correctly enforce behavior at the point at which the test is written, but as you point out, **this is not the same as proving that the code itself works in the real world at all**.

In fact, there is an argument which exists that the test itself is a form of technical debt. 

From the moment that the test is written and the code is released, there is an ever-increasing possibility that `kubeadm` will change in a way that renders the code in Salt non-functional. 

Not only does the test not do anything to signal whether or not the code actually *works*, but at the moment that `kubeadm` changes and we don't detect it, we have a test that is essentially giving us a false-positive. This is very bad.

While I do agree that the test might provide value for another developer, if the test can't be guaranteed to accurately represent the underlying service with which Salt is interacting, how is giving the second developer an unreliable view into the health of the code they are modifiying a benefit?

It is for this reason that is why I urge the community to have a conversation about what requirements might be necessary to build out a test suite in which code put under test can actually be run such that integration tests can be easily constructed and tested against the versions of the downstream software in the real world.

If we can achieve this goal then I'm completely in agreement with you. Until, then, I'm afraid we might just have to disagree. :) If this policy goes forward and developers respond by writing fundamentally unreliable unit tests instead of integration tests (which they will pretty much have to at present), then the project is accumulating technical debt -- not removing it.

Reply:
> > I cannot prove that kubeadm will do the right thing there, because I am not the owner of kubeadm in the code. Is outside the domain of the module and the unit tests that I am providing.
> 
> Right. Here is the heart of my concern. I agree that a unit test of an execution module might correctly enforce behavior at the point at which the test is written, but as you point out, **this is not the same as proving that the code itself works in the real world at all**.

This is a bit unfair.

Those tests are proving something. Is not the full spectrum of what can be done, as integration tests can do. But I am testing the *interface* between salt and kubeadm.

If we have more tests like than, we can change code from one architecture without breaking the code for others.

Of course integration tests needs to be in place. They are relevant and critical for a project like salt.

But testing the interface is a minimal guarantee that needs to be provided as those units tests prevent breakage. Still is much more than no testing anything at all.

Also note that running the unit tests is more time efficient that running the integration tests. If a unit test detect a breakage, you save time.

> From the moment that the test is written and the code is released, there is an ever-increasing possibility that `kubeadm` will change in a way that renders the code in Salt non-functional.

And this is fair, and the code and the tests can be changed.

When a future developer find that this is the case, she / he can make two choices: update the code and replace the tests, or better, check the version of kubeadm and provide the new parameters / behavior for the new kubeadm version and maintain the old behavior and parameters for older versions of kubeadm. The good thing is that she / he will have already tests that guarantee that the old one is still not broken. Can you provide the same guarantee without providing tests?

I am not arguing that tests are a pacanea, and that will fix all our problems. But each PR that provide a test will improve the code a bit, but any PR without a PR is potentially breaking the code and worsening the overall quality of the project.

> Not only does the test not do anything to signal whether or not the code actually _works_, but at the moment that `kubeadm` changes and we don't detect it, we have a test that is essentially giving us a false-positive. This is very bad.

I agree that the tests are not validating the full path: from the user call to the side-effects of the kubeadm action. But this is not what my code is doing, as is only the interface. And this is properly tested.

Will you feel more comfortable I a drop the unit tests for this PR? Will you trust more in the project removing those test / AKA technical debs?

> While I do agree that the test might provide value for another developer, if the test can't be guaranteed to accurately represent the underlying service with which Salt is interacting, how is giving the second developer an unreliable view into the health of the code they are modifiying a benefit?

Because the test is not guaranteeing this. Is about the interface between salt and the external service.

The tests can be broken, and so my code, but each one is confronting the other, and meanwhile none of them is validating the actions of kubeadm, they validate how to call kubeadm as the user expect to.

But again, I return you the question. How an untested code will give guarantee to a developer that is changing the code that is not breaking the old behavior?

> It is for this reason that is why I urge the community to have a conversation about what requirements might be necessary to build out a test suite in which code put under test can actually be run such that integration tests can be easily constructed and tested against the versions of the downstream software in the real world.

I very much agree in any direction that will make easy the integration tests.

But still not see how this transform all the unit tests as a technical deb.

> If we can achieve this goal then I'm completely in agreement with you. Until, then, I'm afraid we might just have to disagree. :) If this policy goes forward and developers respond by writing fundamentally unreliable unit tests instead of integration tests (which they will pretty much have to at present), then the project is accumulating technical debt -- not removing it.

If the tests are useless, lets point that during the review and help the developer to improve them. If the reviewers fails on pointing that, yes, we have a problem.

Reply:
> But I am testing the interface between salt and kubeadm.

You're testing the integration between Salt and a _partial representation_ of that interface which has been specially constructed for the purposes of the Salt module under test. You are *not* testing the interface itself. This is the key point.

To test the interface itself, we would need the software running in an integration test. As we've discussed, while that representation might be accurate, it also might not be. As time goes on, the chances of it being inaccurate increases. 

I understand that you're making the case that if the representation of the interface is not accurately represented that this should be called out in the review. I agree, certainly. 

However, this only solves the case at the time the code is merged, which is not my main concern. My concern is what happens at the point that the software which Salt is controlling diverges from the representation that we're testing against.

This is why I am choosing to call it out as a specific form of technical debt in which the debt becomes due at the point at which a test may be incorrectly reporting a working module when it turns out that the module is not functional.

There are at least two things we could do to improve this situation:

1) We could do a better job of documenting and calling out what version(s) of underlying software is supported by Salt's execution module layer. If we can explicitly say that a given module or integration is only tested up to the version whose interface is represented in the test(s), that goes a long way toward easing my concern. Even better might be considering a policy where modules do not work with untested versions of downstream by default and enabling this capability would involve setting a configuration flag in Salt.

In my view, such a change would do *dramatically more* to improve the stability of Salt in day-to-day use than requiring tests everywhere would.

2) We should forge a clear roadmap for using kitchen-salt (or SUSE's "toaster" project) for figuring out a real path toward integration testing. If a developer could, for example, package up version(s) of software to be used in an integration test and have it run in the test suite, this would be a much larger win than simply enforcing tests everywhere when there is not a test infrastructure right now capable of adequately supporting this pattern.

Reply:
> > But I am testing the interface between salt and kubeadm.
> 
> You're testing the integration between Salt and a _partial representation_ of that interface which has been specially constructed for the purposes of the Salt module under test. You are _not_ testing the interface itself. This is the key point.

I fail to understand this, as I am testing (in the sense that this is where the `assert`s are in the tests) the interface between Salt and the CLI. This is the same interface that the user have.

This kind of test are really common out there:

* https://github.com/intel/govmm/blob/master/qemu/qemu_test.go#L97
* https://github.com/libvirt/libvirt/blob/master/tests/qemuargv2xmldata/disk-drive-network-iscsi.args
* ...

[...]
> However, this only solves the case at the time the code is merged, which is not my main concern. My concern is what happens at the point that the software which Salt is controlling diverges from the representation that we're testing against.

I understand this. I make a point about that. But let me repeat: if kubeadm changes do you want to maintain retro-compatibility? Not all the users will have the last and best kubeadm available. So I would say that yes, we want that. What guarantee this retro-compatibility? I argue that are the test. When a developer add the support of the new kubeadm, the test will protect of breaking the old behavior.

Is the same for multi-architecture CLI tools. GNU vs BSD user space, etc.

Following your argument there is not value in testing the CLI calls to the mount command, as are different in Linux, BSD, Solaris and AIX. You will call this effort technical debt, as a test for `mount` is Linux is useless in AIX, for example. Happens that a change in the way that mount works on AIX broke the call in all Linux system. Do we really want this?


> There are at least two things we could do to improve this situation:
> 
>     1. We could do a better job of documenting and calling out what version(s) of underlying software is supported by Salt's execution module layer. 
[...]

> In my view, such a change would do _dramatically more_ to improve the stability of Salt in day-to-day use than requiring tests everywhere would.

I agree that documenting this is very relevant.

I do not agree that is a replacement. I make the same point again and again: PR without tests will break the old-good behavior. There is not documentation to replace this.

>     1. We should forge a clear roadmap for using kitchen-salt (or SUSE's "toaster" project) for figuring out a real path toward integration testing. If a developer could, for example, package up version(s) of software to be used in an integration test and have it run in the test suite, this would be a much larger win than simply enforcing tests everywhere when there is not a test infrastructure right now capable of adequately supporting this pattern.

I agree totally that improving integration tests is a must.

I totally disagree that this is a valid replacement for unit test. This is a complement on top of them. Will improve the quality as they will find more errors, but you still requires the unit tests to avoid breakages and make it easy for the developer to validate the new changes.

I think that we can end the discussion here. Is OK to disagree and I did my best to point why the requirement of tests is a must.

Of course I am totally in favor of improving the documentation and making a better history for integration tests, but I cannot agree that those are any kind of replacement of testing what the new PR is about.

Reply:
>>This is why I am choosing to call it out as a specific form of technical debt in which the debt becomes due at the point at which a test may be incorrectly reporting a working module when it turns out that the module is not functional from the perspective of the user.

>I'm not sure that this is different at all from not having tests. The benefit of having tests, however, is that when frobnosticate upgrades from version 3 to 4, now we have tests for version 3, so when we make updates for version 4 and we inadvertently break code for version 3 we have a quick feedback showing that we did, in fact, break existing behavior.

It is honestly troubling that you see no distinction between the two. I'm not being hyperbolic here.

Scenario 1: Test exists which isn't actually testing the module, but rather the _other logic in the function_, with external dependencies mocked. The test passes because we've controlled for the operation of the module. Does it work? We don't know for sure, but the passing test lets us _think_ that it is working.

Scenario 2: No test. Does it work? We have no idea.

These are not remotely the same. Sure, a unit test where you've mocked out the thing you're testing helps keep you from breaking logic with later changes to Salt, and that's not nothing. But it's also not a substitute for properly testing the operation of the module, and it doesn't help you to detect upstream changes. So when the upstream module changes, you'd better hope it does so in a way that causes the mocks you wrote to no longer allow the test to pass, or else you're oblivious to a change that is meanwhile causing the module to break for the people actually using that module.

>The other end of that spectrum either shifts the burden to "testing in production", i.e. Salt's users, or a manual process of testing on multiple versions.

The first half of this sentence isn't any different than rolling the dice with unit tests. We're still waiting for upstream code to diverge, cause Salt to stop working, and hope someone files a bug report. (But at least we know we're not breaking the module for older versions, so... yay?)

The second is just nonsensical, and I don't know where it even comes from, so I don't have a response for it. :man_shrugging: 

>So yes, there's a cost to maintaining the tests, but it's not like the alternative is free. There's a cost there, too.

I don't think that anyone is arguing that having no tests isn't costly, so let's dispense with this straw-man argument, please.


Unit tests are just one part of a test infrastructure, arguably a minor part, and the fact that they're easier to write will make them the go-to if we have a dogmatic policy to require tests instead of making the focus what it _should_ be: to come up with the best way to accurately test how the software actually works.

I'd recommend watching this excellent talk from this year's PyCon: https://www.youtube.com/watch?v=Ldlz4V-UCFw

Reply:
> > > This is why I am choosing to call it out as a specific form of technical debt in which the debt becomes due at the point at which a test may be incorrectly reporting a working module when it turns out that the module is not functional from the perspective of the user.
> 
> > I'm not sure that this is different at all from not having tests. The benefit of having tests, however, is that when frobnosticate upgrades from version 3 to 4, now we have tests for version 3, so when we make updates for version 4 and we inadvertently break code for version 3 we have a quick feedback showing that we did, in fact, break existing behavior.
> 
> It is honestly troubling that you see no distinction between the two. I'm not being hyperbolic here.
> 
> Scenario 1: Test exists which isn't actually testing the module, but rather the _other logic in the function_, with external dependencies mocked. The test passes because we've controlled for the operation of the module. Does it work? We don't know for sure, but the passing test lets us _think_ that it is working.
> 
> Scenario 2: No test. Does it work? We have no idea.

When I say that I'm unclear on the distinction, what I mean is that I don't see why technical debt enters into the discussion.

The reason I say that is because the discussion is around this question:

What happens when system outside of Salt changes in a breaking way?

- No tests
  - Salt breaks
- Only tests that aren't an accurate reflection of the underlying system
  - Salt breaks

So why does technical debt about *those* tests enter the conversation? That's what I don't understand. Tests that don't use real 3rd party systems do not tell us anything about 3rd party systems. Is anyone arguing that they do?

> These are not remotely the same. Sure, a unit test where you've mocked out the thing you're testing helps keep you from breaking logic with later changes to Salt, and that's not nothing. But it's also not a substitute for properly testing the operation of the module, and it doesn't help you to detect upstream changes.

Right. I *think* that's what everyone is saying. 

> So when the upstream module changes, you'd better hope it does so in a way that causes the mocks you wrote to no longer allow the test to pass, or else you're oblivious to a change that is meanwhile causing the module to break for the people actually using that module.

I don't think anyone is disagreeing with that statement either?

> > So yes, there's a cost to maintaining the tests, but it's not like the alternative is free. There's a cost there, too.
> 
> I don't think that anyone is arguing that having no tests isn't costly, so let's dispense with this straw-man argument, please.

What *are* we arguing about anyway? I just looked back through this discussion... ah. Okay, now I see. The question is about this statement:

> Test Coverage: tests written to cover new feature.

And the question is "what good will this do?

The answer then, should be this:

- Provide an increased confidence in the code in question
  - With accompanying tests, at least there is enough understanding of what the code *should* be doing that automated tests can be written
  - Spending some extra time writing tests *may* indicate more thought/care went into the design.
- Unit tests provide increased confidence that code does not change in an unexpectedly backwards-incompatible fashion
- Functional tests provide increased confidence that salt interactions with the underlying systems behave as expected 
- End-to-end/Integration tests provide increased confidence that Salt as a whole operates correctly.

> Unit tests are just one part of a test infrastructure, arguably a minor part

Agreed.

>, and the fact that they're easier to write will make them the go-to if we have a dogmatic policy to require tests instead of making the focus what it _should_ be: to come up with the best way to accurately test how the software actually works.

🤷‍♂ That may be correct. But is it better to have *no* tests, or *some* tests? Because what we're talking about is the difference between merging code that has no tests, and merging code that has at least *some* tests. Or changing this SEP to be more explicit about test requirements.

My personal opinion is that currently the way it's stated is fine. My current experience says that *some* tests are better than *no* tests. And not spelling out exactly what kinds of tests we're requiring gives the community more flexibility - just want some unit tests? Cool, that's a great start. Want to create an entire set of tests, with unit, functional, and integration tests, as well as extensive documentation and examples? Awesome!

But that's just *my* strong opinion, weakly held.

> I'd recommend watching this excellent talk from this year's PyCon: https://www.youtube.com/watch?v=Ldlz4V-UCFw

Nice! That definitely illustrates a lot of the challenges to proper testing, and I've definitely seen those problems myself. Hopefully created them less often 😝 I know that I definitely have a preference for the stub-style doubles and DI/functional style code/tests myself.

I know that we will probably get some PRs that have tests that mock too much, and some that will commit other anti-patterns. My hope is that we'll catch these during the PRs and can give people guidance on how to write better tests. I think this is a worthwhile first step that will produce a lot of good information that we can use as we work to improve our test infrastructure. I think it's a Goldilocks requirement - it's not too small and it's not too big, it's *just* right.

Reply:
>The reason I say that is because the discussion is around this question:
>
>What happens when system outside of Salt changes in a breaking way?
>
>- No tests
>    - Salt breaks
>- Only tests that aren't an accurate reflection of the underlying system
>    - Salt breaks
>
>So why does technical debt about those tests enter the conversation? That's what I don't understand. Tests that don't use real 3rd party systems do not tell us anything about 3rd party systems. Is anyone arguing that they do?

No. For the umpteenth time, the concern (and an entirely valid one) is that this will lead to unit tests being over-represented due to their expediency. The technical debt comes from waiting to write better tests, or just _not writing them at all_.

Exhibit A:
```
% find tests/unit -type f -name test_\*.py | wc -l
726
% find tests/integration -type f -name test_\*.py | wc -l
224
```

This ratio is not going to get better if we mandate that tests are written without making it easier to write more accurate tests.

>My current experience says that some tests are better than no tests.

This is reductive to the point of absurdity.

>I know that we will probably get some PRs that have tests that mock too much, and some that will commit other anti-patterns. My hope is that we'll catch these during the PRs and can give people guidance on how to write better tests. I think this is a worthwhile first step that will produce a lot of good information that we can use as we work to improve our test infrastructure.

"Hope" is a particularly bad strategy on which to base the future of the test infra. You can "hope" all you like that we're going to use unit tests as a starting point and build around them, but as someone who is far more familiar with Salt and how we've approached testing over the years, I can say that this is not very likely. The tests that are written all too frequently just become "the tests" for that module, unless repeated bug reports cause us to write better ones.

A hard requirement like this is for _after_ we've approached creating better ways to test complex things. Not before.

Reply:
> This represents a substantial barrier to entry for new feature development. It does not seem to me to be a responsible policy to implement until such time as their is at least some basic testing in place for all modules. Is there a plan to achieve this at present?

>I see below that there is an exemption for this policy which means that that PR is left open for three months but it's not clear to me what good that does. Could somebody please clarify?

Just want to go back to your original questions surrounding what happens when a PR is left open @cachedout 

I do think that we should more clearly define how we would approach the situation where a user is not able to write a test and we do not want to merge without test coverage. One idea was to run community salt-sprints where we add tests for these specific PRs, but I think its fair to say that we should plan these out so they occur often, and are not just wishful thinking. These test focused salt-sprints could also slowly try to tackle this concern:

>"If you just look at unit test coverage for modules right now, this condition wherein a module exists but no unit test file is present would apply in just under half the cases. (Somewhere near 43% with rough math.)"  

What ya think of this approach? I'm sure you will point out something I'm missing ;)

Reply:
This could work, my concern is that how do you make sure that the subject matter expert is in charge of making the tests for the module?

I could definitely see people writing tests for things that they don't understand and then all that is being done is writing tests for coverage purposes instead of for actually testing the important bits.

Reply:
Great question. Maybe we ensure certain test coverage is assigned out to the appropriate work groups during the sprint? Although this only covers certain areas for now as we have only started up a couple work groups this time around. 

I'm not sure how to ensure someone is a subject matter expert unless I'm familiar with them in the community, just the same as i do not know if they are an expert when they push a pull request. The only other thought I have is to make it clear in the announcement beforehand that these are the tests we are covering and the areas and if you have expertise in this area please join to help write tests. Also to make it clear the intention was also to include team-core in this sprint to help tackle these test issues as well so they should be able to help in some of those areas that others do not have expertise in. 

I think another thing that will help is for our team to triage the PRs/tests beforehand and we can add a bit of direction on each PR on how someone could add adequate test coverage.